src/parser.rs

```diff
@@ -1480,6 +1482,13 @@ fn collect_definitions(seg: &mut Segment) {
         if stmt.stype.takes_label() {
             seg.labels.push(LabelDef { index });
         }
+        if let Axiom | Provable = stmt.stype {
+            seg.outline.push(HeadingDef {
+                name: (*b"").into(),
+                index,
+                level: HeadingLevel::Statement,
+            });
+        }
```

Comment:
Isn't this going to significantly increase the size of the outline data structure? These statements all exist in a contiguous block, so ideally it would be possible to only store the position of the start and end, and then you can fake an extra fifth level of the outline in the iterators.

Reply:
Yes - I've checked just now, `HeadingDef` is 24 bytes, and more than 40.000 are stored, so that's 1MB gone.

Metamath-knife stores everything in `Statements`, including also constant declarations, comments, disjoint variable declarations, opening and closing blocks, so statements are not in a contiguous block, but I could skip those statements we are not interested in when iterating through them. I'll try that out.

Reply:
It's not so easy: statements and actual chapters might be mixed (like in the picture below in Alexander's Mathbox)
![image](https://user-images.githubusercontent.com/5831830/149808961-b66fbbff-e433-4a10-899e-f5124c17a1f1.png)

Reply:
I think that means that statements can appear at any level in the tree, so every outline item needs to track its statement address, and the iterator must be prepared to yield statements between every outline item.

Reply:
This is now done in the latest commit.

With that commit, the `OutlineNodeRef` is an enum which either points to a chapter in the outline tree, or an actual statement, using a statement ref. Iterating through children of such a node will first iterate through the statements which are right below the chapter (right after the chapter heading), then through the sub-chapters.